staticmethod tests don't work with fixtures

Original post by @ceridwen in [this comment](https://github.com/pytest-dev/pytest/issues/2528#issuecomment-323212432):

---

I was attempting to use this, and it doesn't seem to be working quite right with fixtures.  Running the following code,

```python
import pytest

@pytest.fixture
def foo():
    return 'foo'

def test_foo(foo):
    assert foo == 'foo'

class TestFoo:
    @staticmethod
    def test_foo(foo):
        assert foo == 'foo'
```

Produces the following error:

```
pyfuncitem = <Function 'test_foo'>

    @hookimpl(trylast=True)
    def pytest_pyfunc_call(pyfuncitem):
        testfunction = pyfuncitem.obj
        if pyfuncitem._isyieldedfunction():
            testfunction(*pyfuncitem._args)
        else:
            funcargs = pyfuncitem.funcargs
            testargs = {}
            for arg in pyfuncitem._fixtureinfo.argnames:
                testargs[arg] = funcargs[arg]
>           testfunction(**testargs)
E           TypeError: test_foo() missing 1 required positional argument: 'foo'

../../Library/Python/3.6/lib/python/site-packages/_pytest/python.py:143: TypeError
======================= 1 failed, 1 passed in 0.32 seconds =======================
```

The first `test_foo()` function passes correctly.

--- a/_pytest/compat.py
+++ b/_pytest/compat.py
@@ -83,6 +83,11 @@ def num_mock_patch_args(function):
 
 
 def getfuncargnames(function, startindex=None, cls=None):
+    """
+    @RonnyPfannschmidt: This function should be refactored when we revisit fixtures. The
+    fixture mechanism should ask the node for the fixture names, and not try to obtain
+    directly from the function object well after collection has occurred.
+    """
     if startindex is None and cls is not None:
         is_staticmethod = isinstance(cls.__dict__.get(function.__name__, None), staticmethod)
         startindex = 0 if is_staticmethod else 1

--- a/_pytest/compat.py
+++ b/_pytest/compat.py
@@ -82,7 +82,10 @@ def num_mock_patch_args(function):
     return len(patchings)
 
 
-def getfuncargnames(function, startindex=None):
+def getfuncargnames(function, startindex=None, cls=None):
+    if startindex is None and cls is not None:
+        is_staticmethod = isinstance(cls.__dict__.get(function.__name__, None), staticmethod)
+        startindex = 0 if is_staticmethod else 1
     # XXX merge with main.py's varnames
     # assert not isclass(function)
     realfunction = function

--- a/_pytest/fixtures.py
+++ b/_pytest/fixtures.py
@@ -957,11 +957,7 @@ class FixtureManager:
 
     def getfixtureinfo(self, node, func, cls, funcargs=True):
         if funcargs and not hasattr(node, "nofuncargs"):
-            if cls is not None:
-                startindex = 1
-            else:
-                startindex = None
-            argnames = getfuncargnames(func, startindex)
+            argnames = getfuncargnames(func, cls=cls)
         else:
             argnames = ()
         usefixtures = getattr(func, "usefixtures", None)

--- a/changelog/2699.bugfix
+++ b/changelog/2699.bugfix
@@ -1,0 +1,1 @@
+Allow tests declared as ``@staticmethod`` to use fixtures.

--- a/testing/python/collect.py
+++ b/testing/python/collect.py
@@ -147,11 +147,21 @@ class TestClass(object):
         ])
 
     def test_static_method(self, testdir):
+        """Support for collecting staticmethod tests (#2528, #2699)"""
         testdir.getmodulecol("""
+            import pytest
             class Test(object):
                 @staticmethod
                 def test_something():
                     pass
+
+                @pytest.fixture
+                def fix(self):
+                    return 1
+
+                @staticmethod
+                def test_fix(fix):
+                    assert fix == 1
         """)
         result = testdir.runpytest()
         if sys.version_info < (2, 7):
@@ -162,8 +172,8 @@ class TestClass(object):
             ])
         else:
             result.stdout.fnmatch_lines([
-                "*collected 1 item*",
-                "*1 passed in*",
+                "*collected 2 items*",
+                "*2 passed in*",
             ])
 
     def test_setup_teardown_class_as_classmethod(self, testdir):

--- a/testing/python/fixture.py
+++ b/testing/python/fixture.py
@@ -29,9 +29,15 @@ def test_getfuncargnames():
         def f(self, arg1, arg2="hello"):
             pass
 
+        @staticmethod
+        def static(arg1, arg2):
+            pass
+
     assert fixtures.getfuncargnames(A().f) == ('arg1',)
     if sys.version_info < (3, 0):
         assert fixtures.getfuncargnames(A.f) == ('arg1',)
+
+    assert fixtures.getfuncargnames(A.static, cls=A) == ('arg1', 'arg2')
 
 
 class TestFillFixtures(object):

